spec/index.bs

```diff
@@ -13055,6 +12999,78 @@ this incurs a clear of the alpha channel. If a canvas is only needed for interop
+
+    1. If |context|.{{GPUCanvasContext/[[currentTexture]]}} is not `null`:
+        1. Call |context|.{{GPUCanvasContext/[[currentTexture]]}}.{{GPUTexture/destroy()}}
+            (without destroying |context|.{{GPUCanvasContext/[[drawingBuffer]]}})
+            to terminate write access to the image.
+        1. Set |context|.{{GPUCanvasContext/[[currentTexture]]}} to `null`.
+</div>
+
+## HTML Specification Hooks ## {#canvas-hooks}
+
+The following algorithms "hook" into algorithms in the HTML specification, and must run at the
+specified points.
+
+<div algorithm="get the bitmap of a WebGPU canvas">
+    When the "bitmap" is read from an {{HTMLCanvasElement}} or {{OffscreenCanvas}} with a
+    {{GPUCanvasContext}} |context|:
+
+    1. Return [$get a copy of the image contents of a context|a copy of the image contents$]
+        of |context|.
+
+    <div class=note>
+        Note:
+        This occurs in many places, including:
+
+        - When an {{HTMLCanvasElement}} has its rendering updated.
+        - When an {{OffscreenCanvas}} with a [=placeholder canvas element=] has its rendering updated.
+        - When {{OffscreenCanvas/transferToImageBitmap()}} creates an {{ImageBitmap}} from the bitmap.
+        - When WebGPU canvas contents are read using other Web APIs, like
+            {{CanvasDrawImage/drawImage()}}, `texImage2D()`, `texSubImage2D()`,
+            {{HTMLCanvasElement/toDataURL()}}, {{HTMLCanvasElement/toBlob()}}, and so on.
+
+        If {{GPUCanvasConfiguration/alphaMode}} is {{GPUCanvasAlphaMode/"opaque"}},
+        this incurs a clear of the alpha channel. Implementations may skip this step when
+        they are able to read or display images in a way that ignores the alpha channel.
+
+        If an application needs a canvas only for interop (not presentation), avoid
+        {{GPUCanvasAlphaMode/"opaque"}} if it is not needed.
+    </div>
+</div>
+
+<div algorithm>
+    When <dfn abstract-op>updating the rendering of a WebGPU canvas</dfn>
+    (an {{HTMLCanvasElement}} or an {{OffscreenCanvas}} with a [=placeholder canvas element=])
+    with a {{GPUCanvasContext}} |context|, which occurs in the following sub-steps of the
+    [=event loop processing model=]:
+
+    - "update the rendering or user interface of that `Document`"
+    - "update the rendering of that dedicated worker"
+
+    Run the following steps:
+
+    1. [$Expire the current texture$] of |context|.
+
+        Note: If this already happened in the task queued by
+        {{GPUCanvasContext/getCurrentTexture()}}, it has no effect.
+
+    Note:
+    This does not happen for standalone {{OffscreenCanvas}}es (created by `new OffscreenCanvas()`).
+</div>
+
+<div algorithm="transferToImageBitmap from WebGPU">
+    When {{OffscreenCanvas/transferToImageBitmap()}} is called on a canvas with
+    {{GPUCanvasContext}} |context|, after creating an {{ImageBitmap}} from the canvas's bitmap:
+
+    1. [$Replace the drawing buffer$] of |context|.
```

Comment:
I'm not sure if I understand why this calls the replace algorithm, while the others call the expire algorithm? The replace algorithm doesn't update the buffer of the current texture in place or anything like that, so I don't see the harm in waiting till the next `getCurrentTexture()` is called to replace the buffer?

Reply:
Actually, I take that back. I understand that we want to have a new drawing buffer in place for the next time the canvas gets composite, if applicable. I'm a bit concerned that if we look at the algorithm for `getCurrentTexture()`, though, it also replaces the drawing buffer. This means that if I'm using, for example, an offscreen canvas and doing a `transferToImageBitmap()` every frame as my presentation method a strict reading of this text would have me allocating two drawing buffers per frame. Once after the transfer, and once during `getCurrentTexture()`.

Maybe we could either have this method replace the `[[currentTexture]]`? Or have it remove the drawing buffer without immediately replacing, then have `get the bitmap of a WebGPU canvas` step allocate one just-in-time if needed?

Reply:
> I understand that we want to have a new drawing buffer in place for the next time the canvas gets composite, if applicable.

It also affects readback operations (texImage2D etc). The prose explanation of the design is in the definitions of [[drawingBuffer]] and [[currentTexture]]. https://gpuweb.github.io/gpuweb/#dom-gpucanvascontext-drawingbuffer-slot

> would have me allocating two drawing buffers per frame

True per spec, but the one created in transferToImageBitmap can't have been modified (and typically won't even have been read) before getCurrentTexture, so one of the two can be elided. I'll add a note about it.
(Also note, this isn't a change relative to the current spec.)

> Maybe we could either have this method replace the `[[currentTexture]]`? Or have it remove the drawing buffer without immediately replacing, then have `get the bitmap of a WebGPU canvas` step allocate one just-in-time if needed?

It's not strictly necessary to replace the drawing buffer in transferToImageBitmap, as expiring the texture makes it readonly anyway. But it matches the usual semantics of transferToImageBitmap (https://html.spec.whatwg.org/multipage/canvas.html#dom-offscreencanvas-transfertoimagebitmap).

We could write out the lazy semantics of the [[drawingBuffer]] slot in the spec, but I feel like it mostly just adds more complexity. Writing it this way also matches the way transferToImageBitmap is defined.

Reply:
SGTM! A non-normative note is probably the most it needs then.